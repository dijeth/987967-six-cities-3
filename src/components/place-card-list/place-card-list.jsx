import React from 'react';
import PropTypes from 'prop-types';
import PlaceCard from '../place-card/place-card.jsx';
import ActionCreator from '../../reducers/app/action-creator.js';
import {Operation} from '../../reducers/operation.js';
import {connect} from 'react-redux';
import {offerPropType} from '../../const/props.js';
import withActiveItem from '../../hocs/with-active-item/with-active-item.jsx';
import {handleActiveOfferChange} from '../../hocs/with-pathname/with-pathname.jsx';

const PlaceCardList = ({items, isNearPlaces, onOfferHover, onListClick}) => {
  const classList = isNearPlaces ? `near-places__list places__list` : `cities__places-list places__list tabs__content`;

  const placeCardList = items.map((offer, i) => {
    return (
      <PlaceCard
        offer={offer}
        isNearPlaces={isNearPlaces}
        key={offer.id}
        onHover={isNearPlaces ? null : onOfferHover}
        offsetIndex={i}
      />);
  });

  return (
    <div
      className={classList}
      onClick={onListClick}
    >
      {placeCardList}
    </div>);
};

PlaceCardList.propTypes = {
  items: PropTypes.arrayOf(offerPropType).isRequired,
  isNearPlaces: PropTypes.bool.isRequired,
  onOfferHover: PropTypes.func.isRequired,
  onActiveItemChange: PropTypes.func,
  onListClick: PropTypes.func,
  activeItem: offerPropType
};

const mapDispatchToProps = (dispatch) => ({
  onActiveItemChange(activeItem) {
    handleActiveOfferChange(dispatch, activeItem.id)
  },

  onOfferHover(offer) {
    dispatch(ActionCreator.changeActiveOffer(offer ? offer.id : null));
  }
});

export {PlaceCardList};
export default connect(null, mapDispatchToProps)(withActiveItem(PlaceCardList, `.place-card__name`));
