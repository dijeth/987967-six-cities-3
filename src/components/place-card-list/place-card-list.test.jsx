import React from 'react';
import renderer from 'react-test-renderer';
import {PlaceCardList} from './place-card-list.jsx';
import {BrowserRouter, Switch, Route} from 'react-router-dom';

const mocks = [{
  id: `id-1`,
  title: `title-1`,
  type: `Apartment`,
  pictures: [`picture-1`],
  cost: 123,
  rating: 4.8,
  isPremium: false,
  isFavorite: true,
  city: `Amsterdam`,
  coord: [1, 2]
},
{
  id: `id-2`,
  title: `title-2`,
  type: `Apartment`,
  pictures: [`picture-2`],
  cost: 456,
  rating: 5,
  isPremium: true,
  isFavorite: false,
  city: `Amsterdam`,
  coord: [3, 4]
}
];

describe(`<PlaceCardList /> should be render correctly`, () => {
  it(`when nearPlacesFor === undefined, isAuth === false`, () => {
    const tree = renderer.create(
        <BrowserRouter>
          <PlaceCardList
            items={mocks}
            onOfferHover={()=>{}}
            isAuth={false}
          />
          <Switch>
            <Route exact path="/offer/:id" />
          </Switch>
        </BrowserRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`when nearPlacesFor === undefined, isAuth === true`, () => {
    const tree = renderer.create(
        <BrowserRouter>
          <PlaceCardList
            items={mocks}
            onOfferHover={()=>{}}
            isAuth={true}
          />
          <Switch>
            <Route exact path="/offer/:id" />
          </Switch>
        </BrowserRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`when nearPlacesFor === 1`, () => {
    const tree = renderer.create(
        <BrowserRouter>
          <PlaceCardList
            items={mocks}
            nearPlacesFor={`1`}
            onOfferHover={()=>{}}
            isAuth={true}
          />
          <Switch>
            <Route exact path="/offer/:id" />
          </Switch>
        </BrowserRouter>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
