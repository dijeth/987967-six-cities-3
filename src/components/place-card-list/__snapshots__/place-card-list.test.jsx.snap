// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaceCardList /> should be render correctly 1`] = `
<div
  className="cities__places-list places__list tabs__content"
>
  <article
    className="cities__place-card place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="cities__image-wrapper place-card__image-wrapper"
    >
      <a
        href="#"
      >
        <img
          alt="Place image"
          className="place-card__image"
          height="200"
          src="picture-1"
          width="260"
        />
      </a>
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            123
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button  place-card__bookmark-button--active button"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            To bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "100%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        <a
          href="#"
          onClick={[Function]}
        >
          title-1
        </a>
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
  <article
    className="cities__place-card place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__mark"
    >
      <span>
        Premium
      </span>
    </div>
    <div
      className="cities__image-wrapper place-card__image-wrapper"
    >
      <a
        href="#"
      >
        <img
          alt="Place image"
          className="place-card__image"
          height="200"
          src="picture-2"
          width="260"
        />
      </a>
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            456
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button false button"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            To bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "100%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        <a
          href="#"
          onClick={[Function]}
        >
          title-2
        </a>
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
</div>
`;
